File: ol-objc/ol-objc/OLSourceValue.m
//
//  OLSourceValue.m
//  ol-objc
//
//  Created by 伍威 on 15/1/7.
//  Copyright (c) 2015年 sleepsaint. All rights reserved.
//

#import "OLSourceValue.h"
#import "OLValue.h"
#import "OLSource.h"
#import "OLFunction.h"

@implementation OLSourceValue

- (id) lookup:(id)root temp:(id)temp now:(id)now {
    return nil;
}
- (NSComparisonResult) compare3:(id)root temp:(id)temp a:(id)a b:(id)b {
    return [[self lookup:root temp:temp now:a] compare2:[self lookup:root temp:temp now:b]];
}
@end

id autoLookup(id root, id temp, id now, id current) {
    while ([current isKindOfClass: NSString.class]) {
        OLSourceValue* source = [OLSource parse:current];
        current = [source lookup:root temp:temp now:now];
    }
    return current;
}

@implementation OLString

+ (instancetype) value:(NSString *)string {
    OLString* ret = [[OLString alloc] init];
    ret->_value = string;
    return ret;
}
- (id) lookup:(id)root temp:(id)temp now:(id)now {
    return _value;
}
@end

@implementation OLNumber

+ (instancetype) value:(double)number; {
    OLNumber* ret = [[OLNumber alloc] init];
    ret->_value = [NSNumber numberWithDouble:number];
    return ret;
}
- (id) lookup:(id)root temp:(id)temp now:(id)now {
    return _value;
}

@end


@implementation OLPath

+ (instancetype) root:(int)root keys:(NSArray*)keys {
    OLPath* ret = [[OLPath alloc] init];
    ret->_root = root;
    ret->_keys = keys;
    return ret;
}

- (id) lookup:(id)root temp:(id)temp now:(id)now {
    id current;
    switch (_root) {
        case '^':
            current = root;
            break;
        case '~':
            current = temp;
            break;
        case '@':
            current = now;
            break;
        default:
            return nil;
    }
    for (OLSourceValue* k in _keys) {
        id key = [k lookup:root temp:temp now:now];
        current = autoLookup(root, temp, now, current);
        if (key && current) {
            current = [current getValueByKey:key];
        } else {
            return nil;
        }
    }
    return current;
}
- (NSArray*) sort:(NSArray*)array root:(id)root temp:(id)temp {
    return [array sortedArrayUsingComparator:^(id a, id b){
        return [self compare3:root temp:temp a:a b:b];
    }];
}

@end

@implementation OLList

+ (instancetype) head:(OLSourceValue*)head tail:(NSArray*)tail {
    OLList* ret = [[OLList alloc] init];
    ret->_head = head;
    ret->_tail = tail;
    return ret;
}
- (id) lookup:(id)root temp:(id)temp now:(id)now {
    id name = [_head lookup:root temp:temp now:now];
    if (name) {
        NSMutableArray* params = [NSMutableArray arrayWithCapacity:_tail.count];
        for (OLSourceValue* s in _tail) {
            id value = [s lookup:root temp:temp now:now];
            [params addObject:value?value:[NSNull null]];
        }
        return [OLFunction calc:name params:params root:root temp:temp now:now];
    }
    return nil;
}
- (NSArray*) sort:(NSArray*)array root:(id)root temp:(id)temp {
    return [array sortedArrayUsingComparator:^(id a, id b){
        NSComparisonResult ret = [self->_head compare3:root temp:temp a:a b:b];
        for (OLSourceValue* s in self->_tail) {
            if (ret != NSOrderedSame) {
                return ret;
            } else {
                ret = [s compare3:root temp:temp a:a b:b];
            }
        }
        return ret;
    }];
}
@end

@implementation OLNegative

+ (instancetype) value:(OLSourceValue*)value {
    OLNegative* ret = [[OLNegative alloc] init];
    ret->_value = value;
    return ret;
}
- (id) lookup:(id)root temp:(id)temp now:(id)now {
    return [NSNumber numberWithBool:![[_value lookup:root temp:temp now:now] boolValue]];
}
- (NSArray*) sort:(NSArray*)array root:(id)root temp:(id)temp {
    return [array sortedArrayUsingComparator:^(id a, id b){
        return -[self->_value compare3:root temp:temp a:a b:b];
    }];
}

@end

@implementation OLQuote

+ (instancetype) value:(OLSourceValue*)value {
    OLQuote* ret = [[OLQuote alloc] init];
    ret->_value = value;
    return ret;
}
- (id) lookup:(id)root temp:(id)temp now:(id)now {
    return _value;
}

@end

File: ol-swift/ol-swift/OLSourceValue.swift
//
//  OLSourceValue.swift
//  ol-swift
//
//  Created by 伍 威 on 15/1/7.
//  Copyright (c) 2015年 sleepsaint. All rights reserved.
//

import Foundation

class OLSourceValue: OLValue {
    func lookup(root: OLValue, temp: OLValue, now: OLValue) -> OLValue {
        return NSNull()
    }
    func toBool() -> Bool {
        return true
    }
    func getValueByKey(key: OLValue) -> OLValue {
        return NSNull()
    }
    func compare(b: OLValue) -> Int {
        return 0
    }
    func filter(function: OLList, root: OLValue, temp: OLValue) -> OLValue {
        return self
    }
    func some(function: OLList, root: OLValue, temp: OLValue) -> Bool {
        return true
    }
    func toArray() -> [AnyObject] {
        return []
    }
    func sort(inout array: [AnyObject], root: OLValue, temp: OLValue) {
        
    }
    func compare(a: OLValue, _ b: OLValue, root: OLValue, temp: OLValue) -> Int {
        return lookup(root, temp: temp, now: a).compare(lookup(root, temp: temp, now: b))
    }
}

class OLString: OLSourceValue, Printable {
    let value: NSString
    init(value: NSString) {
        self.value = value
    }
    var description : String {
        return value
    }
    override func lookup(root: OLValue, temp: OLValue, now: OLValue) -> OLValue {
        return value
    }
}

class OLNumber: OLSourceValue, Printable {
    let value: NSNumber
    init(value: NSNumber) {
        self.value = value
    }
    var description : String {
        return "\(value)"
    }
    override func lookup(root: OLValue, temp: OLValue, now: OLValue) -> OLValue {
        return value
    }
}

class OLPath: OLSourceValue, Printable {
    let root: Character
    var keys = Array<OLSourceValue>()
    init(root: Character) {
        self.root = root
    }
    func addKey(key: OLSourceValue) {
        keys.append(key)
    }
    var description : String {
        return "\(root)\(keys)"
    }
    override func lookup(root: OLValue, temp: OLValue, now: OLValue) -> OLValue {
        var current : OLValue
        switch self.root {
        case "^":
            current = root
        case "~":
            current = temp
        case "@":
            current = now
        default:
            return NSNull()
        }
        for k in keys {
            let key = k.lookup(root, temp: temp, now: now)
            current = autoLookup(root, temp, now, current)
            current = current.getValueByKey(key)
        }
        return current
    }
    override func sort(inout array: [AnyObject], root: OLValue, temp: OLValue) {
        array.sort({
            return self.compare(toOLValue($0), toOLValue($1), root: root, temp: temp) < 0
        })
    }
}

class OLList: OLSourceValue, Printable {
    let head: OLSourceValue
    var tail = Array<OLSourceValue>()
    init(head: OLSourceValue) {
        self.head = head
    }
    func addItem(item: OLSourceValue) {
        tail.append(item)
    }
    var description : String {
        return "\(head)(\(tail))"
    }
    override func lookup(root: OLValue, temp: OLValue, now: OLValue) -> OLValue {
        if let name = head.lookup(root, temp: temp, now: now) as? NSString {
            if let function = OLFunction[name] {
                return function(tail.map({$0.lookup(root, temp: temp, now: now)}), root, temp, now)
            }
        }
        
        return NSNull()
    }
    override func sort(inout array: [AnyObject], root: OLValue, temp: OLValue) {
        array.sort({
            var ret = self.head.compare(toOLValue($0), toOLValue($1), root: root, temp: temp)
            for i in self.tail {
                if ret != 0 { return ret < 0 }
                else { ret = i.compare(toOLValue($0), toOLValue($1), root: root, temp: temp) }
            }
            return ret < 0
        })
    }
}

class OLNegative: OLSourceValue, Printable {
    let value: OLSourceValue
    init(value: OLSourceValue) {
        self.value = value
    }
    var description : String {
        return "!\(value)"
    }
    override func lookup(root: OLValue, temp: OLValue, now: OLValue) -> OLValue {
        return true
    }
    override func sort(inout array: [AnyObject], root: OLValue, temp: OLValue) {
        array.sort({
            return self.value.compare(toOLValue($0), toOLValue($1), root: root, temp: temp) > 0
        })
    }
    override func compare(a: OLValue, _ b: OLValue, root: OLValue, temp: OLValue) -> Int {
        return -value.lookup(root, temp: temp, now: a).compare(value.lookup(root, temp: temp, now: b))
    }
}

class OLQuote: OLSourceValue, Printable {
    let value: OLSourceValue
    init(value: OLSourceValue) {
        self.value = value
    }
    var description : String {
        return "#\(value)"
    }
    override func lookup(root: OLValue, temp: OLValue, now: OLValue) -> OLValue {
        return value
    }
}

func autoLookup(root: OLValue, temp: OLValue, now: OLValue, current: OLValue) -> OLValue {
    var ret = current
    while let source = ret as? NSString {
        if let value = OLSource.parse(source) {
            ret = value.lookup(root, temp: temp, now: now)
        }
    }
    return ret
}

